sort/src/batched_vec.rs

```diff
@@ -0,0 +1,150 @@
+use std::{cmp, mem, ptr};
+use stash::Stash;
+
+pub struct BatchedVecRef<'a, T: 'a> {
+    tail: &'a mut Vec<T>,
+    batches: &'a mut Vec<Vec<T>>
+}
+
+impl<'a, T> BatchedVecRef<'a, T> {
+    #[inline(always)]
+    pub fn is_empty(&self) -> bool {
+        // It is sufficent to only check the tail for emptiness, since any time we flush
+        // the tail (in .reserve), we also push some elements.
+        self.tail.is_empty()
+    }
+
+    #[inline(always)]
+    fn reserve(&mut self, stash: &mut Stash<T>) {
+        if self.tail.len() == self.tail.capacity() {
+            let complete = mem::replace(self.tail, stash.get());
+            if !complete.is_empty() {
+                self.batches.push(complete);
+            }
+        }
+    }
+
+    #[inline(always)]
+    pub fn push(&mut self, element: T, stash: &mut Stash<T>) {
+        self.reserve(stash);
+
+        // The reserve call already ensured we have space. For
+        // efficiency, use an unchecked push.
+        unsafe {
+            let len = self.tail.len();
+            ptr::write(self.tail.get_unchecked_mut(len), element);
+            self.tail.set_len(len + 1);
+        }
+    }
+
+    pub fn push_vec(&mut self, mut elements: Vec<T>, stash: &mut Stash<T>) {
+        // Fill up tail so that we keep the batches fully filled (we don't
+        // want to waste memory).
+        let available = self.tail.capacity() - self.tail.len();
+        let to_move = cmp::min(available, elements.len());
+        self.tail.extend(elements.drain(..to_move)); // This ought to compile to just 2 memcpys
+
+        if elements.is_empty() {
+            stash.give(elements);
+        } else {
+            let tail = mem::replace(self.tail, elements);
+            if !tail.is_empty() {
+                self.batches.push(tail);
+            }
+        }
+    }
+
+    #[inline(always)]
+    pub unsafe fn push_all(&mut self, elements: &[T], stash: &mut Stash<T>) {
+        self.reserve(stash);
+
+        if !(self.tail.capacity() - self.tail.len() >= elements.len()) {
+            panic!("cap: {:?}, len: {:?}, pcl: {:?}", self.tail.capacity(), self.tail.len(), elements.len());
+        }
+
+        let len = self.tail.len();
+        ptr::copy_nonoverlapping(elements.as_ptr(), self.tail.as_mut_ptr().offset(len as isize), elements.len());
+        self.tail.set_len(len + elements.len());
+    }
+
+    #[inline(always)]
+    pub fn finish_into(&mut self, target: &mut Vec<Vec<T>>, stash: &mut Stash<T>) {
+        target.extend(self.batches.drain(..));
+        if !self.tail.is_empty() {
+            target.push(mem::replace(&mut self.tail, stash.get()));
```

Comment:
I think in this sort of situation, we might want to avoid `stash.get()` and rather leave behind an unallocated `Vec::new()`, under the reasoning that while we are not using the sorter it shouldn't sit on allocations (other than those in `stash`, which we can manage with `rebalance`). Instead, we can wait until we receive elements the next time around, which should grab an allocated buffer when it sees a capacity zero buffer.